sensors/baro/ms5611.c

```diff
@@ -0,0 +1,280 @@
+		}
+
+		/* Translate press */
+		off = ((int64_t)ctx->prom[2] << 16) + (((int64_t)dT * (int64_t)ctx->prom[4]) >> 7);  /* OFF = C2 * 2^16 + (C4 * dT) / 2^7 */
+		sens = ((int64_t)ctx->prom[1] << 15) + (((int64_t)dT * (int64_t)ctx->prom[3]) >> 8); /* SENS = C1 * 2^15 + (C4 * dT) / 2^8 */
+		press = ((((int64_t)D1 * sens) >> 21) - off) >> 15;                                  /* P = ((D1 * SENS) / 2^21 - OFF) / 2^15 */
+
+		/* MS5611 operating ranges: temperature from -40C to +80C, and pressure: 45000Pa, 110000 Pa */
+		if (press > 45000 && press < 120000 && temp > -4000 && temp < 8500) {
+			ctx->evtBaro.timestamp = now;
+			ctx->evtBaro.baro.pressure = press;
+			ctx->evtBaro.baro.temp = (temp / 100) + 273;
+
+			sensors_publish(info->id, &ctx->evtBaro);
+		}
+	}
+
+	return;
+}
+
+
+static int ms5611_hwSetup(ms5611_ctx_t *ctx)
+{
+	uint8_t cmd, i;
+
+	/* Reset ms5611 sequence */
+	cmd = CMD_RESET;
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), NULL, 0, 0) < 0) {
+		fprintf(stderr, "ms5611: failed to reset device\n");
+		return -1;
+	}
+	usleep(RESET_SLEEP);
+
+	/* PROM reading */
+	for (i = 0; i < PROM_SIZE; i++) {
+		cmd = CMD_READ_PROM | (i << 1); /* LSB of prom address is always 0 */
+		if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &ctx->prom[i], sizeof(ctx->prom[i]), sizeof(cmd)) < 0) {
+			fprintf(stderr, "ms5611: failed read PROM C%i\n", i + 1);
+			return -1;
+		}
+		ctx->prom[i] = (ctx->prom[i] << 8) | (ctx->prom[i] >> 8); /* swap higher and lower bytes */
+	}
+
+	/* Check CRC of PROM (lower nibble of last prom byte) with calculated CRC */
+	if (ms5611_crc4(ctx->prom) != (ctx->prom[PROM_SIZE - 1] & 0x000f)) {
+		fprintf(stderr, "ms5611: wrong PROM crc\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int ms5611_start(sensor_info_t *info)
+{
+	int err;
+	ms5611_ctx_t *ctx = (ms5611_ctx_t *)info->ctx;
+
+	err = beginthread(ms5611_publishthr, 4, ctx->stack, sizeof(ctx->stack), info);
+	if (err >= 0) {
+		printf("ms5611: launched sensor\n");
+	}
+
+	return err;
+}
+
+
+static int ms5611_alloc(sensor_info_t *info, const char *args)
+{
+	ms5611_ctx_t *ctx;
+	char *ss;
+	int err;
+
+	/* sensor context allocation */
+	ctx = malloc(sizeof(ms5611_ctx_t));
+	if (ctx == NULL) {
+		return -ENOMEM;
+	}
+
+	ctx->evtBaro.type = SENSOR_TYPE_BARO;
+	ctx->evtBaro.baro.devId = info->id;
+
+	/* filling sensor info structure */
+	info->ctx = ctx;
+	info->types = SENSOR_TYPE_BARO;
+
+	ctx->spiCtx.mode = SPI_MODE3;
+	ctx->spiCtx.speed = 10000000;
+
+	ss = strchr(args, ':');
+	if (ss != NULL) {
+		*(ss++) = '\0';
+	}
+
+	/* initialize SPI device communication */
+	err = sensorsspi_open(args, ss, &ctx->spiCtx.oid, &ctx->spiSS);
+	if (err < 0) {
+		printf("ms5611: Can`t initialize SPI device\n");
+		free(ctx);
+		return err;
+	}
+
+	/* hardware setup of Barometer */
+	if (ms5611_hwSetup(ctx) < 0) {
+		printf("ms5611: failed to setup device\n");
+		free(ctx);
```

Comment:
Perhaps we'd like to close spi here

Reply:
@HBuczynski I think this applies to all sensorhub drivers as well. Should it be corrected in all drivers now (this PR) or in separate PR?

Reply:
This PR focuses on m5611 barometer. General changes in sensor hub should be introduced in a separate PR.